### PR TITLE
Feat/cwu review

### PR DESCRIPTION
**Pull request by bdolor**
This PR closes issue: [DMM-300]

Includes tests? [N]
Updated docs? [N]

Proposed changes:
- adds 'under review' status to CWU 
- includes notifications

### CODE REVIEW

File: src/front-end/typescript/lib/pages/opportunity/code-with-us/lib/index.ts [hunk 1]
```diff
@@ -44,6 +48,7 @@ export function cwuOpportunityStatusToTitleCase(
     case CWUOpportunityStatus.Canceled:
       return "Cancelled"; // Use British spelling for copy.
   }
+  return "";
```
Comment (IanFonzie) [inline src/front-end/typescript/lib/pages/opportunity/code-with-us/lib/index.ts:51]: The switch cases are all covered so this statement unnecessary.

File: src/front-end/typescript/lib/pages/opportunity/code-with-us/lib/index.ts [hunk 2]
```diff
@@ -98,6 +103,7 @@ export function cwuOpportunityStatusToPresentTenseVerb(
     case CWUOpportunityStatus.Draft:
       return "Update";
   }
+  return "";
```
Comment (IanFonzie) [inline src/front-end/typescript/lib/pages/opportunity/code-with-us/lib/index.ts:106]: Same as above, however you need to add a case for CWUOpportunityStatus.UnderReview to render a non-empty string in the toasts.

File: src/front-end/typescript/lib/pages/opportunity/code-with-us/lib/index.ts [hunk 3]
```diff
@@ -116,6 +122,7 @@ export function cwuOpportunityStatusToPastTenseVerb(
     case CWUOpportunityStatus.Draft:
       return "Updated";
   }
+  return "";
```
Comment (IanFonzie) [inline src/front-end/typescript/lib/pages/opportunity/code-with-us/lib/index.ts:125]: Here as well.

File: src/shared/lib/resources/opportunity/code-with-us.ts [hunk 4]
```diff
@@ -26,25 +27,35 @@ export enum CWUOpportunityEvent {
...
 
+/**
+ * User-defined type guard to narrow raw input to a CWUOpportunityStatus.
+ *
+ * @see {@link https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates}
+ *
+ * @param raw - a string value
+ * @returns boolean
+ */
+function isCWUOpportunityStatus(
```
Comment (IanFonzie) [inline src/shared/lib/resources/opportunity/code-with-us.ts:38]: nice.

File: src/front-end/typescript/lib/pages/opportunity/code-with-us/lib/index.ts (hunk 1 above)
Comment (bdolor) [reply src/front-end/typescript/lib/pages/opportunity/code-with-us/lib/index.ts:51]: ran into compiling errors and this was the only way through...though they may have been false negatives.

**Review by IanFonzie — CHANGES_REQUESTED** [review]
Good work; the front-end has a couple of testing blindspots though. There's a few small issues regarding toasts that I mentioned in the other comments, but a bigger issue is that we're still rendering the "Actions" button for government users after the opportunity is published, which differs from SWU.

File: src/migrations/tasks/20231102143211_cwu-opp-under-review.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+  await connection.schema.raw(` \
+    ALTER TABLE "cwuOpportunityStatuses" \
+    ADD CONSTRAINT "cwuOpportunityStatuses_status_check" \
+    CHECK ("status" IN ('${Object.values(CWUOpportunityStatus).join("','")}')) \
+  `);
+  logger.info("Completed modifying CWUOpportunityStatuses table.");
+}
+
+export async function down(connection: Knex): Promise<void> {
```
Comment (IanFonzie) [inline src/migrations/tasks/20231102143211_cwu-opp-under-review.ts:42]: Also, I don't know how opinionated you are on this, can we update under review statuses to a valid status when migrating down, probably draft? I'm not able to easily migrate down without deleting records

**Comment by bdolor** [comment]
> ...but a bigger issue is that we're still rendering the "Actions" button for government users after the opportunity is published, which differs from SWU.

sounds like this is something that wasn't introduced in this PR, but may still need to be addressed?

File: src/migrations/tasks/20231102143211_cwu-opp-under-review.ts (hunk 5 above)
Comment (bdolor) [reply src/migrations/tasks/20231102143211_cwu-opp-under-review.ts:42]: good suggestion: https://github.com/bcgov/digital_marketplace/pull/430/commits/7f3ecefc86c12149ca5e4424dc303f2aceba6987

File: src/front-end/typescript/lib/pages/opportunity/code-with-us/lib/index.ts (hunk 2 above)
Comment (bdolor) [reply src/front-end/typescript/lib/pages/opportunity/code-with-us/lib/index.ts]: taken care of in https://github.com/bcgov/digital_marketplace/pull/430/commits/d6eff27f9821cb4a669aa36c30115a7b995019a9

File: src/front-end/typescript/lib/pages/opportunity/code-with-us/lib/index.ts (hunk 1 above)
Comment (bdolor) [reply src/front-end/typescript/lib/pages/opportunity/code-with-us/lib/index.ts]: taken care of in https://github.com/bcgov/digital_marketplace/pull/430/commits/d6eff27f9821cb4a669aa36c30115a7b995019a9

File: src/front-end/typescript/lib/pages/opportunity/code-with-us/lib/index.ts (hunk 3 above)
Comment (bdolor) [reply src/front-end/typescript/lib/pages/opportunity/code-with-us/lib/index.ts]: taken care of in https://github.com/bcgov/digital_marketplace/pull/430/commits/d6eff27f9821cb4a669aa36c30115a7b995019a9

File: src/back-end/lib/resources/opportunity/code-with-us.ts
Comment (IanFonzie) [inline src/back-end/lib/resources/opportunity/code-with-us.ts:1]: I think we need to add a publish permissions check on line 750

**Review by IanFonzie — CHANGES_REQUESTED** [review]

File: src/back-end/lib/resources/opportunity/code-with-us.ts
Comment (bdolor) [reply src/back-end/lib/resources/opportunity/code-with-us.ts:1]: should be addressed in https://github.com/bcgov/digital_marketplace/pull/430/commits/b6c904c09e60affd043987b34aeb754333a35362

**Review by IanFonzie — APPROVED** [review]
Looks good!

**Pull request merged**